staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/enabled_features/doc.go

```diff
@@ -0,0 +1,36 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +k8s:validation-gen=TypeMeta
+// +k8s:validation-gen-scheme-registry=k8s.io/code-generator/cmd/validation-gen/testscheme.Scheme
+
+// This is a test package.
+package enabled_features
+
+import (
+	"k8s.io/code-generator/cmd/validation-gen/testscheme"
+)
+
+func FeatureChecker(object, oldObject *T1) map[string]bool {
+	return map[string]bool{"F1": true, "F2": false}
+}
+
+var localSchemeBuilder = testscheme.New()
+
+// +k8s:validation-gen-feature-checker=k8s.io/code-generator/cmd/validation-gen/output_tests/enabled_features.FeatureChecker
```

Comment:
I get what you're doing here, I think, but...

If we put this tag on the type definition, then either:
* The function definition is in *this* package (e.g. the public k8s.io/api/core/v1), which I don't think we want.
* The function is a ref to something like pkg/apis/core/v1/ (which is where we ulimately emit this code) which is OK, but a little wonky to someone who would look at this from outside of k/k
* The function is a ref to some other pkg, which also is a little wonky to someone who would look at this from outside of k/k

How much leakage are we OK with?

Reply:
How about if we take an approach where simply pass in the feature opts from the strategy?  For example;

```go
func (podStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
  featureXEnabled := hasFeatureX(old) || utilfeature.DefaultFeatureGate.Enabled(features.FeatureX) 
  opts := map[string]bool {
    FeatureX: featureXEnabled,
  }
  allErrs = append(allErrs, rest.ValidateUpdateDeclaratively(ctx, opts, legacyscheme.Scheme, obj, old)...)
}
```

This does mean that we're not going to be able to 100% eliminate the hand written code to perform validation, but at least it we're limiting it to performing specific tasks.

The alternative would be to somehow make feature 'in use detection' declarative. Worth exploring?

Reply:
I've updated this PR to avoid any more generation and instead, expose options out to validation callers so that each type's strategy can populate the options as needed.